core/src/main/java/com/graphhopper/routing/lm/LandmarkStorage.java

```diff
@@ -483,26 +481,27 @@ private boolean createLandmarksForSubnetwork(final int startNode, final byte[] s
      * This method specifies the polygons which should be used to split the world wide area to improve performance and
      * quality in this scenario.
      */
-    public void setSpatialRuleLookup(SpatialRuleLookup ruleLookup) {
-        this.ruleLookup = ruleLookup;
+    public void setAreaIndex(AreaIndex<SplitArea> areaIndex) {
+        this.areaIndex = areaIndex;
     }
 
     /**
      * This method makes edges crossing the specified border inaccessible to split a bigger area into smaller subnetworks.
      * This is important for the world wide use case to limit the maximum distance and also to detect unreasonable routes faster.
      */
-    protected IntHashSet findBorderEdgeIds(SpatialRuleLookup ruleLookup) {
+    protected IntHashSet findBorderEdgeIds(AreaIndex<SplitArea> areaIndex) {
         AllEdgesIterator allEdgesIterator = graph.getAllEdges();
         IntHashSet inaccessible = new IntHashSet();
         while (allEdgesIterator.next()) {
             int adjNode = allEdgesIterator.getAdjNode();
-            SpatialRuleSet set = ruleLookup.lookupRules(na.getLat(adjNode), na.getLon(adjNode));
-            SpatialRule ruleAdj = set.getRules().isEmpty() ? null : set.getRules().get(0);
+            List<SplitArea> areas = areaIndex.query(na.getLat(adjNode), na.getLon(adjNode));
+            SplitArea areaAdj = areas.isEmpty() ? null : areas.get(0);
 
             int baseNode = allEdgesIterator.getBaseNode();
-            set = ruleLookup.lookupRules(na.getLat(baseNode), na.getLon(baseNode));
-            SpatialRule ruleBase = set.getRules().isEmpty() ? null : set.getRules().get(0);
-            if (ruleAdj != ruleBase) {
+            areas = areaIndex.query(na.getLat(baseNode), na.getLon(baseNode));
+            SplitArea areaBase = areas.isEmpty() ? null : areas.get(0);
+            // todo: what if both areaAdj and areaBase are null?
```

Comment:
We mark only those as inaccessible which don't have the same area. `null` is also some kind of area which describes the unassigned landmass: https://github.com/graphhopper/graphhopper/blob/master/core/src/main/resources/com/graphhopper/routing/lm/map.geo.json

* Europe -> Russia = inaccessible
* Russia -> Kazakhstan(null) = inaccessible
* Kazakhstan(null) -> Uzbekistan(null) = accessible

Reply:
Yes so this is fine.